### PR TITLE
Complete optimization of `setup` command

### PR DESCRIPTION
**Pull request by 0xjei**
- [x] Simplify the workflow by dividing into specific function blocks.
- [x] Move and adapt some core methods (agnostic or specific but reusable) into the `actions` package.
- [x] Allow related cloud functions to return an error (not only logs).
- [x] Better error messages.

### CODE REVIEW

File: packages/actions/src/helpers/utils.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import fs from "fs"
+import { CircuitMetadata } from "../types"
+import { genesisZkeyIndex } from "./constants"
+
+/**
+ * Extract data contained in a logger-generated file containing information extracted from R1CS file read.
+ * @notice useful for extracting metadata circuits contained in the generated file using a logger
```
Comment (ctrlc03) [inline packages/actions/src/helpers/utils.ts:7]: I would make this more clear, something along the lines of: 
"Extract data from a R1CS metadata file generated with a logger and SnarkJS's method r1cs.info()"

File: packages/phase2cli/src/lib/errors.ts [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
+/** Services */
+export const CORE_SERVICES_ERRORS = {
     FIREBASE_DEFAULT_APP_DOUBLE_CONFIG: `Wrong double default configuration for Firebase application`,
     FIREBASE_TOKEN_EXPIRED_REMOVED_PERMISSIONS: `The Github authorization has failed due to lack of association between your account and the CLI`,
     FIREBASE_USER_DISABLED: `The Github account has been suspended by the ceremony coordinator(s), blocking the possibility of contribution. Please, contact them to understand the motivation behind it.`,
     FIREBASE_FAILED_CREDENTIALS_VERIFICATION: `Firebase cannot verify your Github credentials due to network errors. Please, try once again later.`,
     FIREBASE_NETWORK_ERROR: `Unable to reach Firebase due to network erros. Please, try once again later and make sure your Internet connection is stable.`,
     FIREBASE_CEREMONY_NOT_OPENED: `There are no ceremonies opened to contributions`,
-    FIREBASE_CEREMONY_NOT_CLOSED: `There are no ceremonies ready to finalization`
+    FIREBASE_CEREMONY_NOT_CLOSED: `There are no ceremonies ready to finalization`,
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/errors.ts:11]: ready for finalization

File: packages/phase2cli/src/commands/setup.ts [hunk 3]
```diff
@@ -58,15 +68,418 @@ import {
...
+) => {
+    const spinner = customSpinner(`Uploading ${theme.text.bold(completeFilename)} file to ceremony storage...`, `clock`)
+    spinner.start()
+
+    await multiPartUpload(
+        firebaseFunctions,
+        bucketName,
+        storageFilePath,
+        localPathAndFileName,
+        String(process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB),
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/setup.ts:476]: We never check that these were set in the .env

File: packages/phase2cli/src/commands/setup.ts [hunk 4]
```diff
@@ -76,21 +489,15 @@ import {
...
     // Setup command state.
-    let circuitSequencePosition = 1 // The circuit's position for contribution.
-    let sharedCircomVersion: string = ""
-    let sharedCircomCommitHash: string = ""
-    let readyToSummarizeCeremony = false // Boolean flag to check whether the coordinator has finished to add circuits to the ceremony.
-
-    // Circuits.
-    const circuitsInputData: Array<CircuitInputData> = [] // All circuits input data.
-    const circuits: Array<CircuitDocument> = []
+    let circuitsInputData: Array<CircuitInputData> = [] // All circuits interactive data.
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/setup.ts:492]: Why are we defining this first and not directly as a constant on line 536?

File: packages/phase2cli/src/commands/setup.ts [hunk 5]
```diff
@@ -125,174 +532,37 @@ const setup = async () => {
...
-        // @todo need to investigate the behaviour of this info() method with huge circuits (could be a pain).
-        await r1cs.info(r1csCWDFilePath, logger)
+    const spinner = customSpinner(`Summarizing your ceremony...`, "clock")
+    spinner.start()
 
-        await sleep(2000) // Sleep 2s to avoid unexpected termination (file descriptor close).
+    // Extract circuits metadata.
+    for (const circuitInputData of circuitsInputData) {
+        // Read file which contains the circuit metadata.
+        const r1csMetadataFilePath = getMetadataLocalFilePath(`${circuitInputData.prefix}_metadata.log`)
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/setup.ts:547]: use `commonTerms.foldersAndPathsTerms.metadata`

**Review by ctrlc03 — COMMENTED** [review]
Great work. Left a few comments here and there.

File: packages/phase2cli/src/commands/setup.ts [hunk 6]
```diff
@@ -58,15 +68,418 @@ import {
...
+
+/**
+ * Manage the creation of a ceremony file storage bucket.
+ * @param firebaseFunctions <Functions> - the Firebase Cloud Functions instance connected to the current application.
+ * @param ceremonyPrefix <string> - the prefix of the ceremony.
+ * @returns <Promise<string>> - the ceremony bucket name.
+ */
+const handleCeremonyBucketCreation = async (firebaseFunctions: Functions, ceremonyPrefix: string): Promise<string> => {
+    // Compose bucket name using the ceremony prefix.
+    const bucketName = getBucketName(ceremonyPrefix, process.env.CONFIG_CEREMONY_BUCKET_POSTFIX!)
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/setup.ts:437]: We are never checking the existence of this .env var

File: packages/backend/src/functions/storage.ts [hunk 7]
```diff
@@ -12,219 +12,263 @@ import { getSignedUrl } from "@aws-sdk/s3-request-presigner"
...
+        // Prepare input data.
         const { objectKey, bucketName } = data
 
-        // get the firestore database
+        // Get Firestore DB.
         const firestoreDatabase = admin.firestore()
 
-        // need to get the ceremony prefix from the bucket name
+        // Extract ceremony prefix from bucket name.
         const ceremonyPrefix = bucketName.replace(process.env.CONFIG_CEREMONY_BUCKET_POSTFIX!, "")
```
Comment (ctrlc03) [inline packages/backend/src/functions/storage.ts:172]: Change to `process.env.AWS_CEREMONY_BUCKET_POSTFIX!`

File: packages/backend/src/functions/auth.ts [hunk 8]
```diff
@@ -69,17 +84,20 @@ export const processSignUpWithCustomClaims = functions.auth.user().onCreate(asyn
     ) {
         customClaims = { coordinator: true }
 
-        logMsg(`User ${user.uid} identified as coordinator`, MsgType.INFO)
+        printLog(`Authenticated user ${user.uid} has been identified as coordinator`, LogLevel.DEBUG)
     } else {
         customClaims = { participant: true }
 
-        logMsg(`User ${user.uid} identified as participant`, MsgType.INFO)
+        printLog(`Authenticated user ${user.uid} has been identified as coordinator`, LogLevel.DEBUG)
```
Comment (ctrlc03) [inline packages/backend/src/functions/auth.ts:91]: identified as participant

File: packages/backend/src/functions/ceremony.ts [hunk 9]
```diff
@@ -15,10 +15,10 @@ export const startCeremony = functions.pubsub.schedule(`every 30 minutes`).onRun
     // Get ceremonies in `scheduled` state.
     const scheduledCeremoniesQuerySnap = await queryCeremoniesByStateAndDate(CeremonyState.SCHEDULED, "startDate", "<=")
 
-    if (scheduledCeremoniesQuerySnap.empty) logMsg(GENERIC_LOGS.GENLOG_NO_CEREMONIES_READY_TO_BE_OPENED, MsgType.INFO)
+    if (scheduledCeremoniesQuerySnap.empty) printLog(`No ceremonies ready to be opened`, LogLevel.INFO)
```
Comment (ctrlc03) [inline packages/backend/src/functions/ceremony.ts:18]: Maybe we can remove this log and just log when we start a ceremony?

File: packages/backend/src/functions/ceremony.ts [hunk 10]
```diff
@@ -33,10 +33,10 @@ export const stopCeremony = functions.pubsub.schedule(`every 30 minutes`).onRun(
     // Get ceremonies in `running` state.
     const runningCeremoniesQuerySnap = await queryCeremoniesByStateAndDate(CeremonyState.OPENED, "endDate", "<=")
 
-    if (runningCeremoniesQuerySnap.empty) logMsg(GENERIC_LOGS.GENLOG_NO_CEREMONIES_READY_TO_BE_CLOSED, MsgType.INFO)
+    if (runningCeremoniesQuerySnap.empty) printLog(`No ceremonies ready to be closed`, LogLevel.INFO)
```
Comment (ctrlc03) [inline packages/backend/src/functions/ceremony.ts:36]: as above. What do you think?

File: packages/backend/src/functions/auth.ts [hunk 11]
```diff
@@ -2,21 +2,32 @@ import * as functions from "firebase-functions"
...
+ * which uses the Firebase Authentication service.
  */
 export const registerAuthUser = functions.auth.user().onCreate(async (user: UserRecord) => {
     // Get DB.
     const firestore = admin.firestore()
 
     // Get user information.
-    if (!user.uid) logMsg(GENERIC_ERRORS.GENERR_NO_AUTH_USER_FOUND, MsgType.ERROR)
+    if (!user.uid) {
+        const error = SPECIFIC_ERRORS.SE_AUTH_NO_CURRENT_AUTH_USER
```
Comment (ctrlc03) [inline packages/backend/src/functions/auth.ts:23]: we can maybe use `logAndThrowError` directly here?

File: packages/phase2cli/src/commands/setup.ts (hunk 3 above)
Comment (0xjei) [reply packages/phase2cli/src/commands/setup.ts:479]: We are testing it in the `bootstrapCommandExecutionAndServices` method here

https://github.com/quadratic-funding/mpc-phase2-suite/pull/320/files/7eddb5d7f2fbd8bde2e13df7b835eeb2cbafb0b4#diff-87f6660d36e6e6644127c9acbc50a32dd9601993eb65311996ffdd1f2e066ed7L40

File: packages/phase2cli/src/commands/setup.ts (hunk 4 above)
Comment (0xjei) [reply packages/phase2cli/src/commands/setup.ts:495]: done

File: packages/phase2cli/src/commands/setup.ts (hunk 5 above)
Comment (0xjei) [reply packages/phase2cli/src/commands/setup.ts:548]: done

File: packages/phase2cli/src/commands/setup.ts (hunk 6 above)
Comment (0xjei) [reply packages/phase2cli/src/commands/setup.ts:437]: Same as above, check L40-41 in `lib/services.ts`

File: packages/backend/src/lib/utils.ts [hunk 12]
```diff
@@ -256,7 +256,10 @@ export const tempDownloadFromBucket = async (
     })
 
     if (!response.ok)
-        logMsg(`Something went wrong when downloading the file from the bucket: ${response.statusText}`, MsgType.ERROR)
+        printLog(
```
[the remark this answers is not in the document]
Comment (0xjei) [reply packages/backend/src/lib/utils.ts:223]: Done!

File: packages/backend/src/functions/storage.ts (hunk 7 above)
Comment (0xjei) [reply packages/backend/src/functions/storage.ts:172]: Wow, good catch!

File: packages/backend/src/functions/auth.ts (hunk 8 above)
Comment (0xjei) [reply packages/backend/src/functions/auth.ts:91]: done

File: packages/backend/src/functions/ceremony.ts (hunk 9 above)
Comment (0xjei) [reply packages/backend/src/functions/ceremony.ts:18]: Totally missed the refactoring of these two functions. Let me add in the next commit! Thanks for the heads up!

File: packages/backend/src/functions/auth.ts (hunk 11 above)
Comment (0xjei) [reply packages/backend/src/functions/auth.ts:23]: definitely

**Review by ctrlc03 — APPROVED** [review]

**Pull request merged**